listenbrainz_spark/stats/user/artist.py

```diff
@@ -0,0 +1,100 @@
+import time
+from collections import defaultdict
+from datetime import datetime
+
+from listenbrainz_spark.constants import LAST_FM_FOUNDING_YEAR
+from listenbrainz_spark.path import LISTENBRAINZ_DATA_DIRECTORY
+from listenbrainz_spark.stats import adjust_days, adjust_months, run_query
+from listenbrainz_spark.stats.user.utils import filter_listens
+from listenbrainz_spark.utils import get_listens
+
+
+def get_artists(table):
+    """ Get artist information (artist_name, artist_msid etc) for every user
+        ordered by listen count in a particular time range.
+
+        Args:
+            table (str): name of the temporary table.
+            from_date (str): UNIX timestamp of start time, defaults to LAST_FM_FOUNDING_YEAR
+            to_date (str): UNIX timestamp of end time, defaults to NOW
+
+        Returns:
+            artists: A dict of dicts which can be depicted as:
+                    {
+                        'user1': [{
+                            'artist_name': str, 'artist_msid': str, 'artist_mbids': str, 'listen_count': int
+                        }],
+                        'user2' : [{...}]
+                    }
+    """
+
+    t0 = time.time()
+
+    result = run_query("""
+            SELECT user_name
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(artist_name) as cnt
+              FROM {table}
+          GROUP BY user_name
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+          ORDER BY cnt DESC
+            """.format(table=table))
+
+    rows = result.collect()
+    artists = defaultdict(list)
+    for row in rows:
+        artists[row.user_name].append({
+            'artist_name': row['artist_name'],
+            'artist_msid': row['artist_msid'],
+            'artist_mbids': row['artist_mbids'],
+            'listen_count': row['cnt'],
+        })
+    print("Query to calculate artist stats processed in %.2f s" % (time.time() - t0))
+    return artists
+
+
+def get_artists_last_week():
+    from_date = adjust_days(datetime.now(), 7)
+    to_date = datetime.now()
+
+    listens_df = get_listens(from_date, to_date, LISTENBRAINZ_DATA_DIRECTORY)
+    filtered_df = filter_listens(listens_df, from_date, to_date)
+    filtered_df.createOrReplaceTempView('user_artist_last_week')
+
+    return get_artists('user_artist_last_week')
+
+
+def get_artists_last_month():
+    from_date = adjust_months(datetime.now(), 1)
+    to_date = datetime.now()
```

Comment:
instead of using datetime.now here and in the `from_date`, it would be better if we used the timestamp of the latest listen we have on the spark cluster, that way we still calculate something even if the dump import pipeline breaks for some reason. The way it works right now, we will delete old calculated stats for the user and start showing them nothing instead.